app/main/presenters/search_presenters.py

```diff
@@ -72,3 +89,73 @@ def set_filter_states(filter_groups, request):
+    """
+    current_lot_slug = get_lot_from_request(request)
+
+    for lot in lots:
+        lot_selected = (lot['slug'] == current_lot_slug)
+        if lot_selected:
+            lot['selected'] = True
+            categories = category_filter_group['filters'] if category_filter_group else []
+            category_was_selected = _show_category_selection(lot, categories, request)
+            lot['categories'] = categories
+
+        if not lot_selected or category_was_selected:
+            # we need a link back to the lot _without_ a category selected
+            lot['link'] = url_for('.search_services', q=get_keywords_from_request(request), lot=lot['slug'])
+
+
+def _show_category_selection(lot, category_filters, request):
+    """
+    Recursive setting of 'selected' state and adding of links to categories
+    as part of building the lots/categories tree.
+    :param category_filters: set of category filters as previously produced by filters_for_question
+    :param request: request object from which to extract active filters
+    :return: True iff a category from the list was selected
+    """
+    request_filters = get_filters_from_request(request)
+    any_category_selected = False
+    any_descendant_selected = False
+
+    for category in category_filters:
+        category['selected'] = False
+        param_values = request_filters.getlist(
+            category['name'],
+            type=str
+        )
+        directly_selected = (category['value'] in param_values)
+        any_child_selected = _show_category_selection(lot, category.get('children', []), request)
+        if directly_selected or any_child_selected:
+            any_category_selected = True
+            category['selected'] = True
+        if any_child_selected:
+            any_descendant_selected = True
+
+        # As with lots, we want a link to the category - but not if the category is directly selected.
+        # If we're showing it as 'selected' because one of its children is selected, then the link is
+        # useful as a kind of breadcrumb, to return to showing all services in that overall category.
+        if not directly_selected:
+            # The argument in the URL must reflect the 'name' of the filter (i.e. the particular question
+            # name it was derived from - 'serviceTypes' for G7/G8, 'serviceCategories' for G9.
+            url_args = {category['name']: category['value']}
+
+            category['link'] = url_for('.search_services',
+                                       q=get_keywords_from_request(request),
+                                       lot=lot['slug'],
+                                       **url_args)
+
+    # Remove parent categories (except the selected one), when any subcategory is selected - we
+    # we think this may make navigation easier (and it's similar to what Amazon does)
+    if any_descendant_selected:
+        category_filters[:] = (c for c in category_filters if c['selected'])  # in-place filter(!)
```

Comment:
`in-place filter(!)` - The fact that you're having to do this suggests that this function has a weird interface, but I understand you're torn by trying to fit in with existing interfaces.